File: multiplication/src/main/java/learn/mssb/mssbmultiplication/challenge/ChallengeAttemptController.java
package learn.mssb.mssbmultiplication.challenge;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.List;

@Slf4j
@RequiredArgsConstructor
@RestController
@RequestMapping("/attempts")
public class ChallengeAttemptController {

    private final ChallengeService challengeService;

    @PostMapping
    public ResponseEntity<ChallengeAttempt> verifyAttempt(@RequestBody @Valid ChallengeAttemptDto challengeAttemptDto) {
        ChallengeAttempt verifiedAttempt = challengeService.verifyAttempt(challengeAttemptDto);
        log.info("Verified attempt: {}", verifiedAttempt);
        return ResponseEntity.ok(verifiedAttempt);
    }

    @GetMapping
    public ResponseEntity<List<ChallengeAttempt>> getStatistics(@RequestParam("alias") String alias) {
        List<ChallengeAttempt> lastAttempts = challengeService.getStatsForUser(alias);
        return ResponseEntity.ok(lastAttempts);
    }
}

File: multiplicationkt/src/main/kotlin/challenge/ChallengeAttemptController.kt
package learn.mssbkt.multiplicationkt.challenge

import jakarta.validation.Valid
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/attempts")
class ChallengeAttemptController(
    private val challengeService: ChallengeService
) {
    private val log = LoggerFactory.getLogger(this::class.java)

    @PostMapping
    fun postResult(@RequestBody @Valid attemptDTO: ChallengeAttemptDTO): ResponseEntity<ChallengeAttempt> {
        val attempt = challengeService.verifyAttempt(attemptDTO)
        log.info("Verified attempt: {}", attempt)
        return ResponseEntity.ok(attempt)
    }

    @GetMapping
    fun getStatistics(@RequestParam("alias") alias: String): ResponseEntity<List<ChallengeAttempt>> {
        return ResponseEntity.ok(challengeService.getStatsForUser(alias))
    }
}
